### PR TITLE
Fix PublishDepsFilePath setting

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
Restore the semantics of PublishDepsFilePath to be location where the deps.json resides when published.
The PublishDepsFilePath property is used by external projects (ex: asp.net) which rely on this definition.

PublishDepsFilePath is empty (by default) for PublishSingleFile, since the deps.json file is embedde within the single-file bundle.

Add a test case for this property.

### CODE REVIEW

**Review by dougbu — COMMENTED** [review]
Works for me but this is @dsplaisted's parade

**Comment by dsplaisted** [comment]
What would the guidance be for consumers of this property if they are run for a single file app?  This change looks like it fixes ASP.NET because it reverts to the previous behavior for a non-single-file app.  But if there's a target that wants to rewrite the deps file and it should work for single-file apps, what are they supposed to do?

**Comment by dougbu** [comment]
Hmm, @dsplaisted would it work to restore the previous (never-empty) values of `$(PublishDepsFilePath)` i.e. eliminate `$(_DepsFilePath)`?

**Comment by swaroop-sridhar** [comment]
> What would the guidance be for consumers of this property if they are run for a single file app? This change looks like it fixes ASP.NET because it reverts to the previous behavior for a non-single-file app. But if there's a target that wants to rewrite the deps file and it should work for single-file apps, what are they supposed to do?

If the meaning of "PublishDepsFilePath"  is really the "actual published location" of the deps file, there's no such location in single-file bundle -- because the deps.json is embedded within the single-file bundle. Therefore, this property is unset for single-file apps.  This behavior is similar to other location semantics in single-file apps. For example, `Assembly.Location` returns the empty string for any embedded assembly.

If a target wants to modify the deps file, it needs to insert itself between `GeneratePublishDependencyFile` and `GenerateSingleFileBundle` targets.  To help with this, we could expose the `_DepsFilePath` target as `IntermediateDepsFilePath` which can be used in single-file or non-single-file cases.

**Comment by swaroop-sridhar** [comment]
> Hmm, @dsplaisted would it work to restore the previous (never-empty) values of `$(PublishDepsFilePath)` i.e. eliminate `$(_DepsFilePath)`?

I don't think this is a good idea -- because `$(PublishDepsFilePath)` will change meaning from 
"published deps file path" to "intermediate deps file path" for single-file apps.

**Comment by swaroop-sridhar** [comment]
@dsplaisted : Made the `_DepsFilePath` property public, as `IntermediateDepsFilePath`.

**Comment by dougbu** [comment]
From my perspective, the current meaning of `$(PublishDepsFilePath)` is "location of the .deps file after the `GeneratePublishDependencyFile` target runs". I don't think naming consistency across `msbuild` properties (aside from the leading underscore) is a good thing to aim for when dealing with legacy. That ship has sailed. And, a name like `$(IntermediateDepsFilePath)` is a strange one if you want to encourage developers to update a file before it's bundled and copied to the final `publish` or `pack` location.

I vote against requiring the world to use a new property instead of `$(PublishDepsFilePath)`.

**Comment by swaroop-sridhar** [comment]
@dougbu  we only need to use `IntermediateDepsFilePath` if updating the deps.json file for single-file apps. 

I think using `IntermediateDepsFilePath`  will make the fact that the deps.file needs to be updated in the "intermediate" stage before bundling into a single-exe, more evident. But I'm not too stuck on the name to use. 

I'm wary of keeping `PublishDepsFilePath` pointing to the `obj\...\app.deps.json` file for single-file apps, because consumers may try to change it post publish (since the file still exists) and expect it to work.

**Comment by dougbu** [comment]
@swaroop-sridhar understand your wariness. I lean toward avoiding breaking changes and you're leaning toward easily explainable general rules. The reason for my leaning is mainly that nobody reads the general rules. But my thoughts on single-file publish are theoretical -- I don't use that infrastructure. If `$(PublishDepsFilePath)` works like it used to for the non-single-file case, at least aspnetcore will be good.

And, if this also fixes the new NU5118 warnings, all the better.

**Comment by swaroop-sridhar** [comment]
My suggestion is to take this fix, and unblock the SDK update. 
We can then discuss what `PublishDepsFilePath` should look like on single-file apps.

**Comment by ViktorHofer** [comment]
Can we please reach consensus here as this currently blocks both aspnetcore and arcade updating to a newer sdk. For aspnetcore this is even on the critical path for preview4.

**Comment by swaroop-sridhar** [comment]
@dsplaisted  please review.

File: src/Tests/Microsoft.NET.Publish.Tests/PublishDepsFilePathTests.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+            var projectPath = Path.Combine(testAsset.Path, testProject.Name);
+            var targetFramework = testProject.TargetFrameworks;
+
+            var publishDir = GetPropertyValue(projectPath, targetFramework, "PublishDir");
+            var projectDepsFileName = GetPropertyValue(projectPath, targetFramework, "ProjectDepsFileName");
+            var publishDepsFilePath = GetPropertyValue(projectPath, targetFramework, "PublishDepsFilePath");
+            var expectedDepsFilePath = $"{publishDir}{projectDepsFileName}";
+
+            publishDepsFilePath.Equals(expectedDepsFilePath).Should().BeTrue();
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Publish.Tests/PublishDepsFilePathTests.cs:45]: I think it's better to write the assertion like this:

```c#
publishDepsFilePath.Should().Be(expectedDepsFilePath)
```

That gives more diagnosable failure messages.

(This also applies to the other assertions)

File: src/Tests/Microsoft.NET.Publish.Tests/PublishDepsFilePathTests.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        public PublishDepsFilePathTests(ITestOutputHelper log) : base(log)
+        {
+        }
+
+        [Fact]
+        public void PublishDepsFilePathIsSetAsExpectedForNormalApps()
+        {
+            var testProject = SetupProject(singleFile: false);
+            var testAsset = _testAssetsManager.CreateTestProject(testProject);
+            var restoreCommand = new RestoreCommand(Log, testAsset.Path, testProject.Name);
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Publish.Tests/PublishDepsFilePathTests.cs:31]: I think you can remove the explicit restore commands.  The GetValuesCommand should restore by default for you.

**Review by dsplaisted — APPROVED** [review]

**Comment by dsplaisted** [comment]
Looks good, we can take this change.  I did not think it was blocking for taking an updated SDK, I thought that there was a workaround.

**Comment by dougbu** [comment]
> I thought that there was a workaround.

It's possible to create a workaround but I don't have one yet. And, if this is also at the root of the new NU5118 warnings, the only workaround there is to suppress the warnings i.e. to sweep things under the rug.

**Comment by dougbu** [comment]
Thanks @swaroop-sridhar and @dsplaisted. About how much time will likely pass between this merge and a new SDK version that dotnet-install.ps1 / dotnet-install.sh can install?

**Comment by dougbu** [comment]
Better question: Is completion of https://dev.azure.com/dnceng/internal/_build/results?buildId=624059&view=results plus ~15 minutes enough for this fix to be available?

**Comment by dsplaisted** [comment]
@dougbu, no, once that build completes then it needs to get inserted into the installer repo (which is automated, though it sometimes needs manual intervention).  Here's the most recent PR that updated the SDK in installer: https://github.com/dotnet/installer/pull/7351

**Comment by dougbu** [comment]
Ah, then https://dev.azure.com/dnceng/internal/_build/results?buildId=624059 completion (soon) + ~15 minutes for Maestro++ machinations + ~2 hours for a dotnet/installer PR and a https://dev.azure.com/dnceng/internal/_build?definitionId=286 run + another ~15 minutes i.e. likely 2.5 hours from now (best case)?

**Comment by dougbu** [comment]
Never mind, the current pipeline just failed in the (Arcade) Signing validation step: https://dev.azure.com/dnceng/internal/_build/results?buildId=624059&view=logs&j=0a6c679f-72be-5867-422e-acb741a068a3&t=d8e44ddf-8ea2-5518-25e7-4300471a271a&l=20 More waiting ☹️

**Comment by dougbu** [comment]
(@dsplaisted @swaroop-sridhar the 'master' branch of dotnet/sdk started failing after an update from microsoft/msbuild. I don't see this error before https://dev.azure.com/dnceng/internal/_build/results?buildId=622370&view=logs&j=0a6c679f-72be-5867-422e-acb741a068a3&t=d8e44ddf-8ea2-5518-25e7-4300471a271a&l=20

**Pull request merged**